binary_transparency/firmware/cmd/flash_tool/impl/flash_tool.go

```diff
@@ -224,13 +236,65 @@ func verifyAnnotations(c *client.ReadonlyClient, pb api.ProofBundle, fwMeta api.
 		return fmt.Errorf("failed to verify update with map checkpoint: %w", err)
 	}
 
-	// TODO(mhutchinson): Check the inclusion proof and use the values returned by the map.
-	afw, _, err := mc.Aggregation(mcp, pb.InclusionProof.LeafIndex)
+	// Get the aggregation and proof, and then check everything about it.
+	// This is a little paranoid as the inclusion proof is generated client-side.
+	// The pretense here is that there is a trust boundary between the code in this class,
+	// and everything else. In a production system, it is likely that the proof generation
+	// would live elsewhere (e.g. in the OTA packaging process), and the shorter proof
+	// bundle would be provided to the device.
+	afw, ip, err := mc.Aggregation(ctx, mcp.Revision, pb.InclusionProof.LeafIndex)
 	if err != nil {
 		return fmt.Errorf("failed to get map value for %q: %w", pb.InclusionProof.LeafIndex, err)
 	}
+	// 1. Check: the proof is for the correct key.
+	kbs := sha512.Sum512_256([]byte(fmt.Sprintf("summary:%d", pb.InclusionProof.LeafIndex)))
+	if !bytes.Equal(ip.Key, kbs[:]) {
+		return fmt.Errorf("received inclusion proof for key %x but wanted %x", ip.Key, kbs[:])
+	}
+	// 2. Check: the proof is for the correct value.
+	preimage, err := json.Marshal(afw)
+	if err != nil {
+		return fmt.Errorf("failed to marshal aggregation to compute preimage: %w", err)
+	}
+	leafID := tree.NewNodeID2(string(kbs[:]), 256)
+	hasher := coniks.Default
+	expectedCommitment := hasher.HashLeaf(api.MapTreeID, leafID, preimage)
+	if !bytes.Equal(ip.Value, expectedCommitment) {
+		// This could happen if the JSON roundtripping was not stable. If we see that happen,
+		// then we'll need to pass out the raw bytes received from the server and parse into
+		// the struct at a higher level.
+		// It could also happen because the value returned is not actually committed to by the map.
+		return fmt.Errorf("received inclusion proof for value %x but wanted %x", ip.Value, expectedCommitment)
+	}
+	// 3. Check: the inclusion proof evaluates to the map root that we've obtained.
+	// The calculation starts from the leaf, and uses the siblings from the inclusion proof
+	// to generate the root, which is then compared with the map checkpoint.
+	calc := expectedCommitment
+	for pd := hasher.BitLen(); pd > 0; pd-- {
+		sib := ip.Proof[pd-1]
+		stem := leafID.Prefix(uint(pd))
+		if sib == nil {
+			sib = hasher.HashEmpty(api.MapTreeID, stem.Sibling())
+		}
+		left, right := calc, sib
+		if !isLeftChild(stem) {
+			left, right = right, left
+		}
+		calc = hasher.HashChildren(left, right)
+	}
+	if !bytes.Equal(calc, mcp.RootHash) {
+		return fmt.Errorf("inclusion proof calculated root %x but wanted %x", calc, mcp.RootHash)
+	}
```

Comment:
Worth pulling this out into a separate func to keep things small?

Reply:
That was one of the approaches I tried on the journey, but the number of parameters ends up being similar to the number of lines extracted! This needs:
 * The expected leaf commitment (`calc`)
 * `hasher`
 * `ip`
 * `leafID`
 * `mcp.RootHash`

If you still think it's worth it then I'll do it, but it felt messy to me.

Reply:
Your call, don't feel super strongly about it, and if this verification needs to be done by something else at some point in the future I guess it can always be extracted then.